includes/class-wc-stripe-payment-tokens.php

```diff
@@ -250,65 +250,99 @@ public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id,
 			return $tokens;
 		}
 
-		$gateway     = WC_Stripe::get_instance()->get_main_stripe_gateway();
-		$upe_gateway = null;
+		try {
+			$gateway  = new WC_Stripe_UPE_Payment_Gateway();
+			$customer = new WC_Stripe_Customer( $user_id );
 
-		foreach ( $gateway->payment_methods as $payment_gateway ) {
-			if ( $payment_gateway->id === $gateway_id ) {
-				$upe_gateway = $payment_gateway;
-				break;
-			}
-		}
+			// Payment methods that exist in Stripe.
+			$stripe_payment_methods     = [];
+			$stripe_payment_methods_ids = [];
 
-		if ( ! $upe_gateway || ! $upe_gateway->is_reusable() ) {
-			return $tokens;
-		}
+			// List of the types already retrieved to avoid pulling redundant information.
+			$types_retrieved_from_stripe = [];
 
-		$customer       = new WC_Stripe_Customer( $user_id );
-		$current_tokens = [];
+			// IDs of the payment methods that exist locally.
+			$locally_stored_payment_methods_ids = [];
 
-		foreach ( $tokens as $token ) {
-			// Store relevant existing tokens here.
-			// We will use this list to check whether these methods still exist on Stripe's side.
-			$current_tokens[ $token->get_token() ] = $token;
-		}
+			// 1. Check if there's any discrepancy between the locally saved payment methods and those saved on Stripe's side.
+			// 2. If local payment methods are not found on Stripe's side, delete them.
+			// 3. If payment methods are found on Stripe's side but not locally, create them.
+			foreach ( $tokens as $token ) {
+				$token_gateway_id = $token->get_gateway_id();
 
-		try {
-			// If this UPE method uses a different payment method type as a token, we don't want to retrieve tokens for it. ie Bancontact uses SEPA tokens.
-			$payment_methods = $customer->get_payment_methods( $upe_gateway->get_retrievable_type() );
+				// The gateway ID of the token doesn't belong to our gateways.
+				if ( ! in_array( $token_gateway_id, self::UPE_REUSABLE_GATEWAYS, true ) ) {
+					continue;
+				}
 
-			// Prevent unnecessary recursion, WC_Payment_Token::save() ends up calling 'woocommerce_get_customer_payment_tokens' in some cases.
-			remove_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
+				$payment_method_type = $this->get_payment_method_type_from_token( $token );
 
-			foreach ( $payment_methods as $payment_method ) {
-				if ( ! isset( $current_tokens[ $payment_method->id ] ) ) {
-					$payment_method_type = $this->get_original_payment_method_type( $payment_method );
```

Comment:
As this `get_original_payment_method_type` is not used anymore, it can be removed I think.

Reply:
Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2897/commits/207d5673c6959b91dc5541b04131cdb15d1c0159